implemented in_app_persistent_mode

guyio opened the pull request:
Hey, I've just added a new switch to winafl: -fuzz_mode in_app_persistent
it will take the normal target_module and target_function/target_offset and hook it into a pre_loop_start_handler instead the pre_fuzz_handler and post_fuzz_handler as we discussed.

If you disagree with something or has any suggestions to make it even better, let me know

ivanfratric commented on afl-fuzz.c:
```diff
@@ -2456,20 +2481,25 @@ static u8 run_target(char** argv, u32 timeout) {
+	  //a workaround for first cycle in app persistent mode
+	  result = ReadCommandFromPipe();
+  }
+  if (result == 0) 
```
Should we check for (result != 'P' && result != 0) case here. In that case something went horribly wrong and we should probably exit.

guyio replied:
agreed

ivanfratric commented on winafl.c:
```diff
@@ -813,6 +879,18 @@ options_init(client_id_t id, int argc, const char *argv[])
             else
                 NOTIFY(0, "Unknown calling convention, using default value instead.\n");
         }
+		else if (strcmp(token, "-fuzz_mode") == 0) {
```
Rename fuzz_mode to persistence_mode? Would make the argument simpler as well, e.g.

-persistence_mode in_app
or
-persistence_mode native
if that is more descriptive

guyio replied:
done

ivanfratric commented on winafl.c:
```diff
@@ -414,12 +431,56 @@ instrument_edge_coverage(void *drcontext, void *tag, instrlist_t *bb, instr_t *i
+		fuzz_target.iteration++;
+
+		//let server know we are starting a new cycle
+		WriteCommandToPipe('P'); 
```
Could the code (both here and in run_target) be made simpler by not sending 'P' here at all?

ivanfratric replied:
Oh, no, nevermind, please ignore.

guyio commented on afl-fuzz.c:
```diff
@@ -2406,6 +2406,33 @@ DWORD WINAPI watchdog_timer( LPVOID lpParam ) {
+			return 0;
+		}
+		PeekNamedPipe(pipe_handle, NULL, 0, NULL, &num_read, NULL);
+	} while (num_read == 0);
```
I've changed the afl-fuzz pipe to be async so it can support timeouts.

guyio commented on afl-fuzz.c:
```diff
@@ -2456,20 +2481,25 @@ static u8 run_target(char** argv, u32 timeout) {
+  }
+  if (result == 0) 
+  {
+	  //saves us from getting stuck in corner case.
```
yes, the read result will be 0 on case of an odd corner case of process exit and timed out read. 
In my opinion this kind of behavior is not interesting and should not divert the fuzzing flow.
In adobe reader for example, I suspect there is some kind of watchdog that terminates the process after some minutes of non returning WaitForSingleObject

guyio commented:
I've added a readme section.

I'm thinking about instead of adding a new test.cpp file I could just change your test.cpp file to contain a while loop option.

what do you think?

ivanfratric commented:
Thanks!

Regarding testing, If it can be done with minimal changes to test.cpp (you'll probably need a command line option to enable the loop and something to serve as target function) I guess that could work.

pipe_overlapped is an interesting idea. Have you checked if it affects performance in any way?

guyio commented:
Hey, I've added a clause for looping in test.cpp, I've just moved the functionality into a separated target function and from main I'm supporting an option to run the target in a loop or just once and return.

Regarding performance, I've tested both old winafl(with busy sync ReadFile) and this forked version(with overlappedio ReadFile) and didn't see any noticeable difference.
But I was quite happy to see that in_app persistent mode added about 20% performance gain.

The performance results on test.exe:
new_afl -loop: ~1000-1100/sec
new_afl no loop: ~900/sec
old_afl:                ~900/sec

ivanfratric commented:
Cool! What is the command line you used for testing?  Might make sense to include it in the readme as well. And what did you use as the target function? fopen?

guyio commented:
I've used the test_target function as target. I'll add an example cmd to the readme for the test.exe.
fopen for example could actually work as a target function with the in_app persistent approach.

ivanfratric commented:
Cool! I tested this today and everything seems to be working normally, so I'm going to merge now. Thank you for contributing!

FYI I won't release a new binary version today because I'm going on vacation for 2 weeks after today and don't want to accidentally break something. Also, there are several other PRs in the queue so I'll probably bundle all of them.

symeonp commented:
Hey @guyio would you mind elaborating a bit on how to instrument specifically the ```fopen``` function? The command you have on the readme works perfectly (```-target_offset 0x1000```) however I'd like to see that behavior as well. Also, am getting ```"ASSERT FAILURE: c:\projects\winafl\winafl.c:925: (0) (invalid option)``` when issuing the ```-persistence_mode in_app``` parameters, there's something in my command line that winafl doesn't like - however it doesn't make sense cause the parameters look correct, did you ever encounter that?

Thanks!

symeonp commented:
Indeed, however on my target unfortunately this doesn't happen, hence I'm trying to figure out what's happening. Here's the command:

```afl-fuzz.exe -i in -o out -D C:\DRIO7\bin32 -t 20000 -- -coverage_module something.dll -fuzz_iterations 5000 -target_module binary.exe -target_method wmain -nargs 2 -persistence_mode in_app -- binary.exe @@```

I have exported the ```wmain``` function (can also confirm with ```dumpbin```) and tried with the offset parameter, still no luck!

symeonp commented:
With that said, to reproduce it, run the following command with an *incorrect* offset .e.g:
```afl-fuzz.exe -i in -o out -D C:\DRIO7\bin32 -t 20000 -- -coverage_module test.exe -fuzz_iterations 5000 -target_module test.exe -target_offest 0x1023 -nargs 2 -persistence_mode in_app -- test.exe @@ loop```
here for example I'm hitting the ```fopen``` function directly which will never work!

guyio commented:
Did you compile winafl yourself? This mode is only on master and still not on a release.

symeonp commented:
Yes I did this morning, but you can also try grabbing the binary from here: https://ci.appveyor.com/project/yoava333/winafl/build/job/7n486f2akspdyqck/artifacts

symeonp commented:
And again: There's nothing wrong with your test binary, that works perfectly (the 0x1000 offest) - am referring to my own harnesses and tests!

guyio commented:
I'm confused, are you getting an error on an argument? or the setup simply doesnt hit the target function?
If you omit the '-persistence_mode in_app' does it run?

symeonp commented:
Yeah ironically I'm having both issues! 
a) I am getting an error regarding an argument and you can reproduce it by instead of passing ```0x1000``` offset to your test binary, try to use a different one (.e.g ```0x1023```) - you will get that error which shouldn't happen. Am attaching screenshot to make things easier. 
![afl_error](https://user-images.githubusercontent.com/954127/44059781-5bc06d2a-9f4a-11e8-919b-6a1cb53667aa.png)

Also b) indeed I never hit the target function

guyio commented:
I've just compiled master and ran the test with the following:
afl-fuzz.exe -i in -o out -D ...bin32... -t 100+ -- -coverage_module test.exe -fuzz_iterations 5000 -target_module test.exe -target_offset 0x1010 -nargs 2 -persistence_mode in_app -- test.exe @@ loop
 and it is working.

also tried with 0x1023 and it times out gracefully.

masthoon commented:
@guyio your pull request added a new command pipe message 'P' sent in _pre_fuzz_handler_.
Both _afl-showmap.c_ and _afl-tmin.c_ were relying on the first message sent to detect the end of fuzzing (see ReadFile in _run_target_), so it prevents tmin and showmap to work properly (they retrieve the coverage information before the fuzzing starts).

guyio commented:
Oh you are right. I didn't ever use these 2.
The right thing to do will be to refactor the 'winafl' run_target into a single place as of now it's a complete code duplication in all executrices.

I'll issue a simple PR that will address the extra 'P' pipe command for both (should be really easy).

But for future work I think this whole winafl pipe should be a single API used from all different apps
